File: FFT_NS_Solver/MATLAB/Get_Initial_Vorticity_From_Stored_Vector_Field/give_Initial_Vorticity_From_Velocity_Field_VTK_File.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
% FUNCTION: from a u.XXXX.vtk file, we will extract U,V (x-Velocity and
% y-Velocity components), compute the background vorticity, and then return
% it to initialize a simulation.
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

function initial_vorticity = give_Initial_Vorticity_From_Velocity_Field_VTK_File()

%
% Grid Parameters (Make sure to match from FFT_NS_Solver.m)
%
Lx = 1;     % x-Length of Computational Grid
Ly = 1;     % y-Length of Computational Grid
Nx = 256;   % Grid resolution in x-Direction
Ny = 256;   % Grid resolution in y-Direction
dx = Lx/Nx; % Spatial step-size in x
dy = Ly/Ny; % Spatial step-size in y

%
% Get back components of velocity field, U,V, e.g., \vec{u} = (U,V)
% 
% Note: U,V are both matrices
%
simNumsString = '0070';
[U,V] = read_Eulerian_Velocity_Field_vtk(simNumsString);

% Note the u.0070.vtk case was performed at 512x512 resolution, hence we
% will downsample for the velocity field, e.g., for 256x256 choose every
% other point
[lenY,~] = size(U);
downsample = lenY/Ny;
U = U(1:downsample:end,1:downsample:end);
V = V(1:downsample:end,1:downsample:end);


%
% Compute Vorticity from Velocity Components
%
initial_vorticity = compute_Vorticity(U,V,dx,dy);

%
% Transpose the vorticity
%
initial_vorticity = initial_vorticity';



%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
% FUNCTION: computes the VORTICITY from the vector field's components, U and V
%           e.g., vec{u} = (U,V)
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

function vorticity = compute_Vorticity(U,V,dx,dy)

% Get Size of Matrix
[Ny,Nx] = size(U);

% Initialize
Uy = zeros(Ny,Nx);
Vx = Uy;

%
% Find partial derivative with respect to x of vertical component of
% velocity, V
%
for i=1:Nx
    
    if i==1
        Vx(:,i) = ( V(1:end,2) - V(1:end,end) ) / (2*dx);
    elseif i==Nx
        Vx(:,i) = ( V(1:end,1) - V(1:end,i-1) ) / (2*dx);
    else
        Vx(:,i) = ( V(1:end,i+1) - V(1:end,i-1) ) / (2*dx);
    end
end

%
% Find partial derivative with respect to y of horizontal component of
% velocity, U
%
for j=1:Ny
    if j==1
        Uy(j,:) = ( U(2,1:end) - U(end,1:end) ) / (2*dy);
    elseif j==Ny
        Uy(j,:) = ( U(1,1:end) - U(j-1,1:end) ) / (2*dy);
    else
        Uy(j,:) = ( U(j+1,1:end) - U(j-1,1:end) ) / (2*dy);
    end
end


% Take difference to find vorticity ( Curl(vec{u}) in 2D )
vorticity = Vx - Uy;




%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
% FUNCTION: Reads in the velocity data field from .vtk format
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

function [U,V] = read_Eulerian_Velocity_Field_vtk(simNums)

filename = ['u.' num2str(simNums) '.vtk'];  

fileID = fopen(filename);
if ( fileID== -1 )
    error('\nCANNOT OPEN THE FILE!');
end

str = fgets(fileID); %-1 if eof
if ~strcmp( str(3:5),'vtk')
    error('\nNot in proper VTK format');
end

% read in the header info %
str = fgets(fileID);
str = fgets(fileID);
str = fgets(fileID);
str = fgets(fileID);
str = fgets(fileID);

% Store grid info
Nx = sscanf(str,'%*s %f %*f %*s',1);
Ny = sscanf(str,'%*s %*f %f %*s',1);


% bypass lines in header %
str = fgets(fileID);
str = fgets(fileID);
str = fgets(fileID);
str = fgets(fileID);
str = fgets(fileID);
str = fgets(fileID);
str = fgets(fileID);


% get formatting for reading in data from .vtk in fscanf %
strVec = '%f';
for i=2:3*Nx
    strVec = [strVec ' %f'];
end

% read in the vertices %
[e_Data,count] = fscanf(fileID,strVec,3*Nx*Ny);
if count ~= 3*Nx*Ny
   error('Problem reading in Eulerian Data.'); 
end

% reshape the matrix into desired data type %
e_Data = reshape(e_Data, 3, count/3); % Reshape (3*Nx*Nx,1) vector to (Nx*Nx,3) matrix
e_Data = e_Data';                     % Store vertices in new matrix

U = e_Data(:,1);       % Store U data
V = e_Data(:,2);       % Store V data

U = reshape(U,Nx,Ny)';  % Reshape (Nx*Nx,1) matrix to  (Nx,Nx)
V = reshape(V,Nx,Ny)';  % Reshape (Nx*Nx,1) matrix to  (Nx,Nx)
 
fclose(fileID);         % Closes the data file.

clear filename fileID str strVec count analysis_path e_Data Nx;

File: FFT_NS_Solver/Python3/Get_Initial_Vorticity_From_Stored_Vector_Field/give_Initial_Vorticity_From_Velocity_Field_VTK_File.py
###########################################################################
#
# FUNCTION: from a u.XXXX.vtk file, we will extract U,V (x-Velocity and
# y-Velocity components), compute the background vorticity, and then return
# it to initialize a simulation.
#
###########################################################################

import numpy as np
import os

#################################################################################
#
# FUNCTION: Reads in (x,y) positions of the immersed boundary from .vtk format
#          
################################################################################

def read_Eulerian_Velocity_Field_vtk(simNums):
    
    simulation_path = os.getcwd();  # Stores working directory's path        
    velocity_path = 'Get_Initial_Vorticity_From_Stored_Vector_Field'
    os.chdir(velocity_path);        # cd's into vorticity initialization folder

    filename = 'u.' + str(simNums) + '.vtk'

    # Stores grid resolution from .vtk file
    Nx = np.genfromtxt(filename, skip_header=5, usecols=(1),max_rows=1)   
    
    # Stores desired Eulerian data
    e_data_X = np.genfromtxt(filename, skip_header=13, usecols=range(0,3*int(Nx),3),max_rows=int(Nx))
    e_data_Y = np.genfromtxt(filename, skip_header=13, usecols=range(1,3*int(Nx),3),max_rows=int(Nx))

    # Go back to working directory
    os.chdir(simulation_path);     

    return e_data_X, e_data_Y

############################################################################
#
# FUNCTION: computes the VORTICITY from the vector field's components, U and V
#           e.g., vec{u} = (U,V)
#
##############################################################################

def compute_Vorticity(U,V,dx,dy):

    # Get Size of Matrix
    [Ny,Nx] = U.shape

    # Initialize storage matrices
    Uy = np.zeros((Ny,Nx))
    Vx = np.zeros((Ny,Nx))

    #
    # Find partial derivative with respect to x of vertical component of
    # velocity, V
    #
    # Recall '-1' equivalent to MATLAB's 'end'
    #
    for i in range(1,Nx+1):
        
        ii=i-1 # index starting at 0

        if (i==1):
            Vx[0:,ii] = ( V[0:,1] - V[0:,-1] ) / (2*dx)
        elif (i==Nx):
            Vx[0:,ii] = ( V[0:,0] - V[0:,ii-1] ) / (2*dx)
        else:
            Vx[0:,ii] = ( V[0:,ii+1] - V[0:,ii-1] ) / (2*dx)
        

    #
    # Find partial derivative with respect to y of horizontal component of
    # velocity, U
    #
    for j in range(1,Ny+1):

        jj=j-1 # index starting at 0

        if (j==1):
            Uy[jj,0:] = ( U[1,0:] - U[-1,0:] ) / (2*dy)
        elif (j==Ny):
            Uy[jj,0:] = ( U[0,0:] - U[jj-1,0:] ) / (2*dy)
        else:
            Uy[jj,0:] = ( U[jj+1,0:] - U[jj-1,0:] ) / (2*dy)



    # Take difference to find vorticity ( Curl(vec{u}) in 2D )
    vorticity = Vx - Uy

    return vorticity


###########################################################################
#
# FUNCTION: from a u.XXXX.vtk file, we will extract U,V (x-Velocity and
# y-Velocity components), compute the background vorticity, and then return
# it to initialize a simulation.
#
###########################################################################

def give_Initial_Vorticity_From_Velocity_Field_VTK_File():

    #
    # Grid Parameters (Make sure to match from FFT_NS_Solver.m)
    #
    Lx = 1     # x-Length of Computational Grid
    Ly = 1     # y-Length of Computational Grid
    Nx = 256   # Grid resolution in x-Direction
    Ny = 256   # Grid resolution in y-Direction
    dx = Lx/Nx # Spatial step-size in x
    dy = Ly/Ny # Spatial step-size in y


    #
    # Get back components of velocity field, U,V, e.g., \vec{u} = (U,V)
    # 
    # Note: U,V are both matrices
    #
    simNumsString = '0070'
    U_orig,V_orig = read_Eulerian_Velocity_Field_vtk(simNumsString)

    # Note the u.0070.vtk case was performed at 512x512 resolution, hence we
    # will downsample for the velocity field, e.g., for 256x256 choose every
    # other point
    lenY,lenX = U_orig.shape
    
    # Amount to downsample by
    downsample = int(lenY/Ny)

    # Store newly downsampled matrices
    U = U_orig[::downsample,::downsample]
    V = V_orig[::downsample,::downsample]

    #
    # Compute Vorticity from Velocity Components
    #
    initial_vorticity = compute_Vorticity(U,V,dx,dy)

    #
    # Transpose the vorticity
    #
    initial_vorticity = np.transpose(initial_vorticity)

    return initial_vorticity
